internal/provider/volume_data_source.go

```diff
@@ -2,98 +2,111 @@ package provider
+	Size   types.Int64  `tfsdk:"size"`
+	Appid  types.String `tfsdk:"app"`
+	Region types.String `tfsdk:"region"`
 }
 
-func (v volumeDataSourceType) GetSchema(context.Context) (tfsdk.Schema, diag.Diagnostics) {
-	return tfsdk.Schema{
+func (d *volumeDataSourceType) Schema(_ context.Context, _ datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = schema.Schema{
 		MarkdownDescription: "Fly volume resource",
-		Attributes: map[string]tfsdk.Attribute{
-			"id": {
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
 				MarkdownDescription: "ID of volume",
-				Type:                types.StringType,
-				Computed:            true,
-				Optional:            true,
+				Required:            true,
+				Validators: []validator.String{
+					stringvalidator.RegexMatches(
+						regexp.MustCompile(`^vol_[a-z0-9]+$`),
+						"must start with \"vol_\"",
+					),
+				},
 			},
-			"app": {
+			"app": schema.StringAttribute{
 				MarkdownDescription: "Name of app attached to",
 				Required:            true,
-				Type:                types.StringType,
 			},
-			"size": {
+			"size": schema.Int64Attribute{
 				MarkdownDescription: "Size of volume in GB",
-				Required:            true,
-				Type:                types.Int64Type,
+				Computed:            true,
 			},
-			"name": {
+			"name": schema.StringAttribute{
 				MarkdownDescription: "name",
-				Type:                types.StringType,
-				Required:            true,
+				Computed:            true,
 			},
-			"region": {
+			"region": schema.StringAttribute{
 				MarkdownDescription: "region",
-				Type:                types.StringType,
-				Required:            true,
-			},
-			"internalid": {
-				MarkdownDescription: "Internal ID",
-				Type:                types.StringType,
-				Required:            true,
+				Computed:            true,
 			},
 		},
-	}, nil
-}
-
-func (v volumeDataSourceType) NewDataSource(_ context.Context, in tfsdkprovider.Provider) (datasource.DataSource, diag.Diagnostics) {
-	provider, diags := convertProviderType(in)
-
-	return volumeDataSource{
-		provider: provider,
-	}, diags
+	}
 }
 
-func (v volumeDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+func (d *volumeDataSourceType) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
 	var data volumeDataSourceOutput
 
 	diags := req.Config.Get(ctx, &data)
 	resp.Diagnostics.Append(diags...)
 
-	internalId := data.Internalid.Value
-	app := data.Appid.Value
+	// strip leading vol_ off name
+	internalId := data.Id.ValueString()[4:]
+	app := data.Appid.ValueString()
 
-	query, err := graphql.VolumeQuery(context.Background(), *v.provider.client, app, internalId)
+	query, err := graphql.VolumeQuery(context.Background(), *d.client, app, internalId)
 	if err != nil {
-		resp.Diagnostics.AddError("Read: query failed", err.Error())
+		resp.Diagnostics.AddError("Query failed", err.Error())
+	}
```

Comment:
Like another one;

- Better to use the passed ctx
- Safer to return early if err != nil